java/sage/io/RemoteSageFile.java

```diff
@@ -0,0 +1,489 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sage.io;
+
+import sage.Sage;
+
+import java.io.*;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+
+public class RemoteSageFile implements SageFileSource
+{
+  private final String hostname;
+  private final String remoteFilename;
+  private final String transcodeMode;
+  private final boolean readonly;
+
+  private boolean activeFile = false;
+  private long currTotalSize = 0;
+  private long maxRemoteSize = 0;
+
+  private long remoteOffset = 0;
+  private Socket socket = null;
+  private DataOutputStream outStream = null;
+  private DataInputStream inStream = null;
+
+  public RemoteSageFile(String hostname, File file, boolean readonly) throws IOException
+  {
+    this(hostname, file.getPath(), readonly, null);
+  }
+
+  public RemoteSageFile(String hostname, String name, boolean readonly) throws IOException
+  {
+    this(hostname, name, readonly, null);
+  }
+
+  public RemoteSageFile(String hostname, File file, boolean readonly, String transcodeMode) throws IOException
+  {
+    this(hostname, file.getPath(), readonly, transcodeMode);
+  }
+
+  /**
+   *
+   *
+   * @param hostname The hostname of the media server hosting this file.
+   * @param name The full path and file to be opened.
+   * @param readonly Set the channel to be read only.
+   * @param transcodeMode Set the transcode mode.
+   * @throws IOException
+   */
+  public RemoteSageFile(String hostname, String name, boolean readonly, String transcodeMode) throws IOException
+  {
+    this.hostname = hostname;
+    remoteFilename = name;
+    this.transcodeMode = transcodeMode;
+    this.readonly = readonly;
+    connect();
+  }
+
+  private void connect() throws IOException
+  {
+    if (socket != null)
+      return;
+
+    Socket newSocket = new Socket();
+    newSocket.connect(new InetSocketAddress(hostname, 7818));
+    newSocket.setSoTimeout(30000);
+    newSocket.setTcpNoDelay(true);
+    socket = newSocket;
+
+    outStream = new DataOutputStream(new BufferedOutputStream(socket.getOutputStream()));
+    inStream = new DataInputStream(new BufferedInputStream(socket.getInputStream()));
+
+    if (transcodeMode != null && transcodeMode.length() > 0)
+    {
+      outStream.write(("XCODE_SETUP " + transcodeMode + "\r\n").getBytes(Sage.BYTE_CHARSET));
+      outStream.flush();
+      String response = Sage.readLineBytes(inStream);
+      if (!"OK".equals(response))
+        throw new java.io.IOException("Error with remote transcode setup for " + transcodeMode + " of: " + response);
+    }
+
+    outStream.write("OPENW ".getBytes(Sage.BYTE_CHARSET));
```

Comment:
I was curious what you were going to do with the 'readonly' field here. :)  When you open a remote file with the media server...you can open it either in read mode or write mode; but not both at the same time (although the MediaServer could be extended to support such a case easily by opening a RandomAccessFile and getting the channel from there rather than a FileOuptutStream object in the openWriteFile method).

The way this is currently done; any calls to the write method in here will fail because the files are being opened in read mode on the server always.  Since you've got all the other work in here...I'd say just change MediaServer to use a RandomAccessFile's channel instead of FileOutputStream (the uploadStream field); then make this use OPENWRITEW instead of OPENW if readonly is set to false (and prior to that, request the upload key from the server so it'll grant you write access). That *should* have no performance implications.